pilot/pkg/bootstrap/server.go

```diff
@@ -246,25 +249,83 @@ func NewServer(args *PilotArgs, initFuncs ...func(*Server)) (*Server, error) {
 		return nil, err
 	}
 
-	if err := s.initControllers(args); err != nil {
+	// Parse and validate Istiod Address.
+	istiodHost, _, err := e.GetDiscoveryAddress()
+	if err != nil {
 		return nil, err
 	}
 
-	s.XDSServer.InitGenerators(e, args.Namespace)
-
-	// Initialize workloadTrustBundle after CA has been initialized
-	if err := s.initWorkloadTrustBundle(args); err != nil {
+	// Create Istiod certs and setup watches.
+	if err := s.initIstiodCerts(args, string(istiodHost)); err != nil {
 		return nil, err
 	}
 
-	// Parse and validate Istiod Address.
-	istiodHost, _, err := e.GetDiscoveryAddress()
-	if err != nil {
+	// federation support must be initialized before config and service controllers
+	if features.EnableFederation {
+		if s.kubeClient == nil {
```

Comment:
This condition will never be true, because if kube client cannot be initialized, the initialization of server will be interrupted with an error. So I think we can just remove this condition.